libraries/platforms/dromajo-vcs/bsg_manycore_simulator.hpp

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2019, University of Washington All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+// 
+// Redistributions of source code must retain the above copyright notice, this list
+// of conditions and the following disclaimer.
+// 
+// Redistributions in binary form must reproduce the above copyright notice, this
+// list of conditions and the following disclaimer in the documentation and/or
+// other materials provided with the distribution.
+// 
+// Neither the name of the copyright holder nor the names of its contributors may
+// be used to endorse or promote products derived from this software without
+// specific prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// This file implements the SimulationWrapper object.
+
+#ifndef __BSG_MANYCORE_SIMULATOR_HPP
+#define __BSG_MANYCORE_SIMULATOR_HPP
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+
+#include <string>
+
+#include <dromajo_cosim.h>
+#include <dromajo_manycore.h>
+
+#include <bsg_nonsynth_dpi_manycore.hpp>
+#include <bsg_nonsynth_dpi_errno.hpp>
+#include <bsg_nonsynth_dpi_cycle_counter.hpp>
+#include <bsg_nonsynth_dpi_clock_gen.hpp>
+
+#include <bsg_manycore.h>
+#include <bsg_manycore_config.h>
+#include <bsg_manycore_printing.h>
+
+#include <bsg_manycore_regression.h>
+
+// Define invisible library functions
+declare_hb_mc_get_bits
+declare_bsg_printing
+
+#define NUM_DROMAJO_INSTR_PER_TICK 1000
+
+#define MC_ARGS_START_EPA_ADDR 0x0000
+#define MC_ARGS_FINISH_EPA_ADDR 0x00FF
+#define MC_CONFIG_START_EPA_ADDR 0x0100
+#define MC_CONFIG_FINISH_EPA_ADDR 0x01FF
+#define MC_RESET_DONE_EPA_ADDR 0x0200
+#define MC_TX_VACANT_EPA_ADDR 0x0300
+#define MC_FINISH_EPA_ADDR 0xEAD0
+#define MC_TIME_EPA_ADDR 0xEAD4
+#define MC_FAIL_EPA_ADDR 0xEAD8
+#define MC_STDOUT_EPA_ADDR 0xEADC
+#define MC_STDERR_EPA_ADDR 0xEAE0
+#define MC_BRANCH_TRACE_EPA_ADDR 0xEAE4
+#define MC_PRINT_STAT_EPA_ADDR 0xEA0C
```

Comment:
Aren't these defined elsewhere? I know finish and fail are https://github.com/bespoke-silicon-group/bsg_replicant/blob/bigblade/libraries/bsg_manycore_epa.h. 

In general though, since these are not simulator specific, some of these are in the wrong place. They seem platform specific, in which case they should be in bsg_manycore_platform.h

Reply:
I thought bsg_manycore_platform.h was actually platform agnostic? Because there isn't one for every platform?

Also, the platform.h file defines the platform-specific functions that need to be implemented like transmit and receive. Is it a good idea to make them visible to the simulator in this case? Mostly because the simulator shouldn't have to know this...

Reply:
I think there are two issues here, so I'll try to clarify.

Defines that are specific to the manycore shouldn't be there. For example FINISH_EPA to PRINT_STAT_EPA should be in epa.h.

For the other ones, what would you do if you wanted to exchange VCS for verilator and you didn't want to re-write code?

Reply:
All the manycore-only EPAs are added to bsg_manycore_epa.h. The others have been moved to the bp_hb_platform.h header. This should be like the one stop shop for all things related to the platform and specific to BP.